File: cpp/forces/drag.cpp
#include "forces_impl.h"

namespace pylumina {
    void DragForce::apply(Particle& p, double /*dt*/) {
        if (p.is_static) return;
        double speed = p.velocity.magnitude();
        if (speed < 1e-6) return;
        
        double magnitude = 0.5 * rho * cd * area * speed * speed;
        Vector3D force = p.velocity.normalized() * (-magnitude);
        p.apply_force(force);
    }
}

File: src/pylumina/physics/drag.py
from .._core import DragForce

class Drag:
    def __init__(self, cd=0.47, area=0.01, rho=1.225):
        self._force = DragForce(cd, area, rho)
    
    @property
    def core(self):
        return self._force
